File: Data_Structure/Cprogramming/ch4_Linked_list_extend/Doubly_LinkedList.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct DoublyListNodeType{
    int data;
    struct DoublyListNodeType *pLLink;
    struct DoublyListNodeType *pRLink;
}DoublyListNode;

typedef struct DoublyListType{
    int currentCount;
    DoublyListNode headerNode;
}DoublyList;

DoublyList *createDoublyList(){
    DoublyList *pReturn = (DoublyList*)malloc(sizeof(DoublyList));
    if(pReturn != NULL){
        memset(pReturn, 0, sizeof(DoublyList));

        pReturn->headerNode.pLLink = &(pReturn->headerNode);
        pReturn->headerNode.pRLink = &(pReturn->headerNode);
    }

    return pReturn;
}

int getDoublyListData(DoublyList* pList, int position){
    int i = 0;

    DoublyListNode *pCurrentNode = &(pList->headerNode);
    for( i = 0 ; i <= position ; i++){
        pCurrentNode = pCurrentNode->pRLink;
    }

    return pCurrentNode->data;
}

int addDoublyListData(DoublyList* pList, int position, int data){
    int i = 0;
    int ret = 0;

    DoublyListNode *pNewNode = NULL;
    DoublyListNode *pPreNode = NULL;

    pNewNode = (DoublyListNode*)malloc(sizeof(DoublyListNode));

    if(pNewNode != NULL){
        memset(pNewNode, 0, sizeof(DoublyListNode));
        pNewNode->data = data;

        pPreNode = &(pList->headerNode);
        for( i = 0 ; i < position ; i++){
            pPreNode = pPreNode->pRLink;
        }

        pNewNode->pRLink = pPreNode->pRLink;
        pNewNode->pLLink = pPreNode;

        pPreNode->pRLink = pNewNode;
        pNewNode->pRLink->pLLink = pNewNode;

        pList->currentCount++;
    }
    else{
        printf("오류, 메모리 할당 addListData()\n");
        return 1;
    }

    return ret;
}

int removeDoublyListData(DoublyList* pList, int position){
    int i = 0, ret = 0;
    DoublyListNode *pDelNode = NULL;
    DoublyListNode *pPreNode = NULL;

    pPreNode = &(pList->headerNode);
    for( i = 0 ; i < position ; i++){
        pPreNode = pPreNode->pRLink;
    }

    pDelNode = pPreNode->pRLink;
    pPreNode->pRLink = pDelNode->pRLink;

    pDelNode->pRLink->pLLink = pPreNode;

    pList->currentCount--;
    free(pDelNode);

    return ret;
}

void deleteDoublyList(DoublyList* pList){
    int i = 0;

    if(pList != NULL){
        while(pList->currentCount > 0){
            removeDoublyListData(pList, 0);
        }
        free(pList);
    }
}

int getDoublyListLength(DoublyList* pList){
    int ret = 0;
    if(NULL != pList){
        return pList->currentCount;
    }
    return ret;
}

void displayDoublyList(DoublyList* pList){
    int i = 0;
    DoublyListNode* pNode = NULL;

    pNode = pList->headerNode.pRLink;
    if( 0 == pList->currentCount ){
        printf("자료가 없습니다\n");
    }
    else{
        printf("노드 개수 : %d\n", pList->currentCount);

        for( i = 0 ; i < pList->currentCount ; i++){
            printf("[%d] = %d\n", i, pNode->data);
            pNode = pNode->pRLink;
        }
    }
}

int main(int argc, char *argv[]){
    DoublyList *pList = NULL;

    pList = createDoublyList();

    addDoublyListData(pList, 0, 10);
    displayDoublyList(pList);
    addDoublyListData(pList, 0, 20);
    displayDoublyList(pList);
    addDoublyListData(pList, 1, 30);
    displayDoublyList(pList);

    removeDoublyListData(pList, 2);
    displayDoublyList(pList);
    removeDoublyListData(pList, 1);
    displayDoublyList(pList);
    removeDoublyListData(pList, 0);
    displayDoublyList(pList);

    deleteDoublyList(pList);

    return 0;
}

File: Data_Structure/python/ch4_LinkedList_extend/Doubly_LinkedList.py
class DoublyListNode:
    def DoublyListNode(self):
        self.data = 0
        self.pLink = {'HeaderClass':0, 'PreClass':0, 'CurrentClass': 0, 'NextClass':0 }
        self.pLink['CurrentClass'] = self
        self.currentCount = 0

def DoublyList(Node):
    DoublyListNode()
    Node.pLink['HeaderClass'] = Node
    return Node

def createDoublyList(Node):
    Node = DoublyListNode()
    Node.DoublyListNode()
    Node.data = 'HeaderNode'
    Node.pLink['PreClass'] = Node
    Node.pLink['NextClass'] = Node
    DoublyList(Node)
    return Node

def getDoublyListData(pList, position):
    i = 0

    if position > pList.currentCount :
        print('\n오류, 불러올 위치 값이 현재 노드 개수 보다 많습니다.\n')
        return

    pCurrentNode = DoublyListNode()
    pCurrentNode.DoublyListNode()
    pCurrentNode = pList.pLink['HeaderClass']

    for i in range(position):
        pCurrentNode = pCurrentNode.pLink['NextClass']

    return pCurrentNode.data

def addDoublyListData(pList, position, data):
    i = 0

    if position == 0 :
        print('\n오류, 값을 넣을 위치는 헤더노드의 위치입니다.\n')
        return

    pNewNode = DoublyListNode()
    pNewNode.DoublyListNode()
    pPreNode = DoublyListNode()
    pPreNode.DoublyListNode()

    pNewNode.data = data

    pPreNode = pList.pLink['HeaderClass']
    for i in range(position-1) :
        pPreNode = pPreNode.pLink['NextClass']

    pNewNode.pLink['NextClass'] = pPreNode.pLink['NextClass']
    pNewNode.pLink['PreClass'] = pPreNode

    pPreNode.pLink['NextClass'] = pNewNode
    pNewNode.pLink['NextClass'].pLink['PreClass'] = pNewNode;


    pList.currentCount += 1

    return pList

def removeDoublyListData(pList, position):
    if position == 0 :
        print('\n오류, 헤더노드는 지울 수 없습니다.\n')
        return

    i = int(0)
    pDelNode = DoublyListNode()
    pDelNode.DoublyListNode()
    pPreNode = DoublyListNode()
    pPreNode.DoublyListNode()

    pPreNode = pList.pLink['HeaderClass']

    for i in range(position - 1):
        pPreNode = pPreNode.pLink['NextClass']

    pDelNode = pPreNode.pLink['NextClass']
    pPreNode.pLink['NextClass'] = pDelNode.pLink['NextClass']

    pDelNode.pLink['NextClass'].pLink['PreClass'] = pPreNode

    pList.currentCount -= 1

def deleteDoublyList(pList):
    if pList.pLink == {} :
        print('\n오류, 리스트가 비어있습니다.\n')
        return

    pDelNode = DoublyListNode()
    pDelNode.DoublyListNode()
    pPreNode = DoublyListNode()
    pPreNode.DoublyListNode()

    pPreNode = pList.pLink['HeaderClass']

    for i in range(pList.currentCount - 1):
        pDelNode = pPreNode
        pPreNode = pPreNode.pLink['NextClass']

        pDelNode.data = 0
        pDelNode.pLink.clear()

def getDoublyListLength(pList):
       return pList.pLink['HeaderClass'].currentCount

def displayList(pList):
    i = 0
    print('\n일반 순회')
    for i in range(pList.currentCount):
        print("Node[{}] = {}".format(i+1,getDoublyListData(pList,i+1)))

def iterateDoublyList(pList):
    count = int(0)

    if pList.pLink == {} :
        print('\n오류, 리스트가 비어있습니다.\n')
        return

    pNode = DoublyListNode()
    pNode.DoublyListNode()

    pNode = pList.pLink['HeaderClass']

    print('\n로직 처리 순회')

    for i in range(pList.currentCount + 1):
        print('{} = {}'.format(count, pNode.data))
        count += 1
        pNode = pNode.pLink['NextClass']

    print('노드 개수 : {}'.format(getDoublyListLength(pList.pLink['HeaderClass'])))

pList = DoublyListNode()
pList.DoublyListNode()
pList = createDoublyList(pList)
addDoublyListData(pList, 1, 10)
addDoublyListData(pList, 2, 20)
addDoublyListData(pList, 2, 30)

value = getDoublyListData(pList, 1)
print('위치 : {}\n값 : {}\n'.format(1,value))

displayList(pList)

iterateDoublyList(pList)

removeDoublyListData(pList, 1)

iterateDoublyList(pList)
deleteDoublyList(pList)